pkg/authorization/authorizer/accessrestriction/accessrestriction.go

```diff
@@ -0,0 +1,215 @@
+		if !matches(accessRestriction, requestAttributes) {
+			continue
+		}
+
+		// it does match, meaning we need to check if it denies the request
+		if a.allowed(accessRestriction, requestAttributes.GetUser()) {
+			glog.V(4).Infof("access restriction %#v matched but did not deny request attributes %#v for user %#v", accessRestriction, requestAttributes, requestAttributes.GetUser())
+			continue
+		}
+
+		// deny the request because it is not allowed by the current access restriction
+		// the reason is opaque because normal users have no visibility into access restriction objects
+		glog.V(2).Infof("access restriction %#v denied request attributes %#v for user %#v", accessRestriction, requestAttributes, requestAttributes.GetUser())
+		return authorizer.DecisionDeny, "denied by access restriction", nil
+	}
+
+	// no access restriction matched and denied this request, so we state that we have no opinion
+	// the reason must be blank, otherwise we would spam all RBAC denies with it (which is generally not useful)
+	return authorizer.DecisionNoOpinion, "", nil
+}
+
+func isIgnored(requestAttributes authorizer.Attributes) bool {
+	// non-resource request is inherently cluster scoped
+	if !requestAttributes.IsResourceRequest() {
+		return true
+	}
+
+	ns := requestAttributes.GetNamespace()
+
+	// is this cluster scoped
+	if ns == v1.NamespaceNone {
+		return true
+	}
+
+	// is this in a reserved namespace
+	return clusterresourceoverride.IsExemptedNamespace(ns)
+}
+
+func matches(accessRestriction *authorizationv1alpha1.AccessRestriction, requestAttributes authorizer.Attributes) bool {
+	if len(accessRestriction.Spec.MatchAttributes) == 0 {
+		return true // fail closed (but validation prevents this)
+	}
+	return rbac.RulesAllow(requestAttributes, accessRestriction.Spec.MatchAttributes...)
+}
+
+func (a *accessRestrictionAuthorizer) allowed(accessRestriction *authorizationv1alpha1.AccessRestriction, user user.Info) bool {
+	return a.subjectsMatch(accessRestriction.Spec.AllowedSubjects, user) || !a.subjectsMatch(accessRestriction.Spec.DeniedSubjects, user)
+}
+
+func (a *accessRestrictionAuthorizer) subjectsMatch(subjects []authorizationv1alpha1.SubjectMatcher, user user.Info) bool {
+	for _, subject := range subjects {
+		if a.subjectMatches(subject, user) {
+			return true
+		}
+	}
+	return false
+}
+
+func (a *accessRestrictionAuthorizer) subjectMatches(subject authorizationv1alpha1.SubjectMatcher, user user.Info) bool {
+	switch {
+	case subject.UserRestriction != nil && subject.GroupRestriction == nil:
+		return a.userMatches(subject.UserRestriction, user)
+	case subject.GroupRestriction != nil && subject.UserRestriction == nil:
+		return a.groupMatches(subject.GroupRestriction, user)
+	}
+	return false // fail closed on whitelist, fail open on blacklist (but validation prevents this)
+}
+
+func (a *accessRestrictionAuthorizer) userMatches(userRestriction *authorizationv1.UserRestriction, user user.Info) bool {
```

Comment:
Refactor the admission logic so you have zero-allocation checks, move it to somewhere more reasonable and re-use it.